### ISSUE TITLE
infer_dtype(df) should return dtype of df values

### ISSUE THREAD

**Issue by auderson** [pandas-dev/pandas#45932]
Currently, `infer_dtype` seems to return dtype of columns of a dataframe

```python
import pandas as pd

df = pd.DataFrame([[.1, .2, .3]], columns=[*'ABC'])

print(pd.api.types.infer_dtype(df))
#  string

df.columns = range(3)
print(pd.api.types.infer_dtype(df))
# integer

print(pd.api.types.infer_dtype(df.values))
# floating
```
This is a bit counter-intuitive. I thought it would work on the values of a dataframe.

**Comment by phofl** [comment]
The input types are defined as
``scalar, list, ndarray, or pandas type``. A DataFrame is neither of that. Maybe we could raise for invalid inputs.

``df.dtype`` returns the dtypes of your DataFrame

**Comment by auderson** [comment]
@phofl 
Yes, I'm aware of that. However, df.dtypes is not useful if the data is `object` type which means it can be datetime or str or other types. In these cases I have to use `infer_dtype`.

**Comment by jreback** [comment]
> The input types are defined as
> 
> ``scalar, list, ndarray, or pandas type``. A DataFrame is neither of that. Maybe we could raise for invalid inputs.
> 
i think this would be fine

**Comment by mroeschke** [comment]
Thanks for the request, but it appears this idea hasn't taken off in a while so closing
